sktime/clustering/k_means/k_means_sklearn.py

```diff
@@ -0,0 +1,253 @@
+        high-dimensional problems (see :ref:`kmeans_sparse_high_dim`).
+
+        When `n_init='auto'`, the number of runs depends on the value of init:
+        10 if using `init='random'` or `init` is a callable;
+        1 if using `init='k-means++'` or `init` is an array-like.
+    max_iter : int, default=300
+        Maximum number of iterations of the k-means algorithm for a
+        single run.
+    tol : float, default=1e-4
+        Relative tolerance with regards to Frobenius norm of the difference
+        in the cluster centers of two consecutive iterations to declare
+        convergence.
+    verbose : int, default=0
+        Verbosity mode.
+    random_state : int, RandomState instance or None, default=None
+        Determines random number generation for centroid initialization. Use
+        an int to make the randomness deterministic.
+        See :term:`Glossary <random_state>`.
+    copy_x : bool, default=True
+        When pre-computing distances it is more numerically accurate to center
+        the data first. If copy_x is True (default), then the original data is
+        not modified. If False, the original data is modified, and put back
+        before the function returns, but small numerical differences may be
+        introduced by subtracting and then adding the data mean. Note that if
+        the original data is not C-contiguous, a copy will be made even if
+        copy_x is False. If the original data is sparse, but not in CSR format,
+        a copy will be made even if copy_x is False.
+    algorithm : {"lloyd", "elkan", "auto", "full"}, default="lloyd"
+        K-means algorithm to use. The classical EM-style algorithm is `"lloyd"`.
+        The `"elkan"` variation can be more efficient on some datasets with
+        well-defined clusters, by using the triangle inequality. However it's
+        more memory intensive due to the allocation of an extra array of shape
+        `(n_samples, n_clusters)`.
+
+    Attributes
+    ----------
+    cluster_centers_ : ndarray of shape (n_clusters, n_features)
+        Coordinates of cluster centers. If the algorithm stops before fully
+        converging (see ``tol`` and ``max_iter``), these will not be
+        consistent with ``labels_``.
+    labels_ : ndarray of shape (n_samples,)
+        Labels of each point
+    """
+
+    _tags = {
+        "capability:multivariate": True,
+        "capability:unequal_length": True,
+        "capability:missing_values": True,
+        "X_inner_mtype": ["pd-multiindex", "numpy3D"],
+        # required by the update_data utility
+        # otherwise, we could pass through to the distance directly
+    }
+
+    DELEGATED_PARAMS = [
+        "n_clusters",
+        "init",
+        "n_init",
+        "max_iter",
+        "tol",
+        "verbose",
+        "random_state",
+        "copy_x",
+        "algorithm",
+    ]
+    DELEGATED_FITTED_PARAMS = ["cluster_centers_", "labels_"]
+
+    def __init__(
+        self,
+        distance,
```

Comment:
@fkiraly, should the `distance` parameter be a string or a callable like the current `metric` parameter in the `TimeSeriesKMeans`?

The docstring for the `TimeSeriesDBSCAN` class (which the `TimeSeriesKMeansSklearn` class follows) says the `distance` parameter can be a string or a callable. However, from what I can tell there is no code to parse strings in the `distance` parameter?

```python
>>> from sktime.clustering.dbscan import TimeSeriesDBSCAN
>>> import numpy as np
>>> x = np.random.normal(size=(50, 2))
>>> model = TimeSeriesDBSCAN(distance="euclidean")
>>> model.fit(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/alex/documents/sktime/sktime/clustering/base.py", line 118, in fit
    self._fit(X)
  File "/home/alex/documents/sktime/sktime/clustering/dbscan.py", line 122, in _fit
    distmat = distance(X)
TypeError: 'str' object is not callable
>>> from sktime.dists_kernels import DtwDist
>>> model = TimeSeriesDBSCAN(distance=DtwDist())
>>> model.fit(x)
TimeSeriesDBSCAN(distance=DtwDist())
```
Am I missing something in my code example above or is the docstring for  `TimeSeriesDBSCAN` slightly out of date?

Reply:
yes, the docstring is incorrect - but it is out of date in the sense that it is the "future desired state" which has not been implemented by accident.

I.e., in the end state, we want all distance based learners to accept strings, and some logic internally converts it to the right callable.

For now, I'd say, you don't need to add that logic because it will be centrally done for all distance based learners - but for correctness' sake, you may want to consider removing the reference to `str` from the docstring until that's implemented.

Reply:
for reference, there's a draft PR here: https://github.com/sktime/sktime/pull/4712
although the end result might look different